test/unit/finance/bill_variable_for_plan_changed_test.rb

```diff
@@ -3,49 +3,56 @@
 require 'test_helper'
 
 class Finance::BillVariableForPlanChangedTest < ActiveSupport::TestCase
-  attr_reader :contract, :account, :app_plan
+  attr_reader :contract, :app_plan
 
   setup do
     @contract = FactoryBot.build_stubbed(:contract)
-    @account  = FactoryBot.build_stubbed(:simple_account)
+    account = FactoryBot.build_stubbed(:simple_account)
     @app_plan = FactoryBot.build_stubbed(:application_plan)
 
-    @org_tz = ENV["TZ"]
+    contract.stubs(:provider_account).returns(account)
+    account.stubs(:provider_can_use?).with(:instant_bill_plan_change).returns(true)
   end
 
-  teardown do
-    ENV["TZ"] = @org_tz
+  test "bill for variable on first day" do
+    Timecop.travel(Time.now.beginning_of_month) # rubocop:disable Rails/TimeZone we don't use timezones in billing
+
+    contract.expects(:save).never
+    contract.notify_observers(:bill_variable_for_plan_changed, app_plan)
   end
 
-  test "bill for variable" do
-    Timecop.travel(15.days.ago) if Time.now.mday == 1 # rubocop:disable Rails/TimeZone we don't use timezones in billing
+  test "bill for variable before first day" do
+    Timecop.travel(Time.now.beginning_of_month - 15.days) # rubocop:disable Rails/TimeZone we don't use timezones in billing
 
-    contract.stubs(:provider_account).returns(account)
-    account.stubs(:provider_can_use?).returns(true)
+    contract.expects(:save).returns(true)
+    contract.notify_observers(:bill_variable_for_plan_changed, app_plan)
+  end
+
+  test "bill for variable after first day" do
+    Timecop.travel(Time.now.beginning_of_month + 15.days) # rubocop:disable Rails/TimeZone we don't use timezones in billing
 
     contract.expects(:save).returns(true)
     contract.notify_observers(:bill_variable_for_plan_changed, app_plan)
   end
 
   test "no variable billing on 1st day of month local time" do
+    org_tz = ENV["TZ"]
+
     # this changes server and client local time to China Standard Time (+8)
     ENV["TZ"] = "Asia/Shanghai"
 
     # this is 4 hours before beginning of month in UTC and 4 after local
     Timecop.travel(Time.zone.now.beginning_of_month - 4.hours)
 
-    contract.stubs(:provider_account).returns(account)
-    account.stubs(:provider_can_use?).returns(true)
-
     contract.expects(:save).never
     contract.notify_observers(:bill_variable_for_plan_changed, app_plan)
+
+    ENV["TZ"] = org_tz
```

Comment:
this should happen in `tierdown` not as part of test code. Otherwise on a failed test we will end up with a changed timezone.